documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_bulk_moderation_actions.md

```diff
@@ -0,0 +1,526 @@
+after it is created. We will add a custom filter to display only bulk moderation
+decisions, defined as a ModerationDecision with more than one media record.
+
+We will disable the option to add a ModerationDecision object directly.
+
+### Report view
+
+No changes are needed to the Report view. Note that by extending the
+ModerationDecision model rather than adding a new model, when a record is viewed
+from the Report view _all_ related moderation decisions (including bulk actions)
+will automatically be visible to moderators.
+
+## Access control
+
+Because bulk moderation decisions can affect a very large number of records,
+access should be limited to only maintainers. We will use Django's existing
+permissions support to ensure that the relevant actions are only visible to
+maintainers.
+
+If we would like to change this policy in the future to allow moderators to make
+bulk decisions, that can be added easily.
+
+## Concurrency
+
+It's possible, even likely, that reports may exist for some records that we'd
+like to take bulk action on. For example, imagine a scenario where a particular
+creator is producing large numbers of sensitive images that are not caught by
+existing filters. We may discover this situation after receiving many user
+reports, and decide to mark all of the creator's records as sensitive in bulk.
+It is therefore necessary to ensure that moderators do not, for example, deindex
+an individual record that is simultaneously being deindexed by a bulk action.
+
+To do this, we will use atomic transactions and Django's
+[select_for_update](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#select-for-update)
+to lock the records being selected for bulk action in each of the custom
+actions.
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+When this work is completed we will be able to do the following new things:
+
+- Filter media in the media views by search terms, or by provider/creator
+- Mark a selection of multiple records as sensitive, as part of a single
+  decision
+- Deindex a selection of multiple records, as part of a single decision
+- Undo marking sensitive and deindexing, for one or more records
+  - This includes undoing individual ModerationDecisions by moderators
+  - This also includes undoing a subset of a previous ModerationDecision
+
+It is equally important to identify what is out of scope for this work:
+
+Performing a bulk moderation decision will only affect records that are
+currently in the API, with no effect on records in the Catalog. For example, if
+a maintainer uses this workflow to deindex all records made by a particular
+Flickr user, this does not prevent the Flickr DAG from ingesting _new_ records
+by that same user, which will be added to the API in the next data refresh.
+Preventing new records from being added, or otherwise propagating deleted
+records to the Catalog, is out of scope for this project.
```

Comment:
Ooo this is an interesting edge case that I had not considered. Excellent to call it out here as out of scope.

Reply:
I wonder if we could in the future add a step to check for media by deindexed creators after the data refresh, and either alert or deindex them all automatically.

Reply:
Definitely! I almost added some ideas to this IP, but it got complicated really quickly 😅 It required adding provider/creator as nullable fields on the ModerationDecision, and added a ton of complexity to the workflows for creating bulk decisions. That also raises questions like: Do we create DeletedMedia/SensitiveMedia records for the new records, or just avoid copying them over in the data refresh? If the former, do we add those to an existing ModerationDecision or create a new one? 🤔 

Ultimately I think it might be best to have some sort of table for tracking creators that should be deindexed and actually prevent ingestion in the catalog. We could have something similar for creators that should be marked sensitive.